### PR TITLE
Display suggestions

### PR DESCRIPTION
**Pull request by Stokalas**
Creating this pull request a bit early just to allow to test calculations.

### CODE REVIEW

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (Stokalas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs]: After deleting Debug code this should have been deleted too.

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 2]
```diff
@@ -8,42 +9,39 @@ namespace ePiggyWeb.DataManagement.Saving
...
-            decimal[] arrayTemplate = new decimal[enumCount];
-            decimal[] sumsOfAmountByImportanceAdjusted = arrayTemplate;
-            decimal[] sumsOfAmountByImportanceDefault = arrayTemplate;
-            decimal[] averagesOfAmountByImportanceAdjusted = arrayTemplate;
-            decimal[] averagesOfAmountByImportanceDefault = arrayTemplate;
+ 
+            decimal[] sumsOfAmountByImportanceAdjusted = new decimal[enumCount];
+            decimal[] sumsOfAmountByImportanceDefault = new decimal[enumCount];
+            decimal[] averagesOfAmountByImportanceAdjusted = new decimal[enumCount];
+            decimal[] averagesOfAmountByImportanceDefault = new decimal[enumCount];
```
Comment (Stokalas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs]: all `decimal[]` should be changed to `var[]`

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 3]
```diff
@@ -54,19 +52,32 @@ public static class AlternativeSavingCalculator
...
             while (goal.Amount > approximateSavedAmount)
             {
                 for (var i = enumCount; i > (int)Importance.Necessary; i--)
                 {
-                    averagesOfAmountByImportanceAdjusted[i - 1] = sumsOfAmountByImportanceAdjusted[i - 1] / entryAmounts[i - 1];
-                    averagesOfAmountByImportanceDefault[i - 1] = sumsOfAmountByImportanceDefault[i - 1] / entryAmounts[i - 1];
+                    if (entryAmounts[i - 1] != 0)
+                    {
+                        averagesOfAmountByImportanceAdjusted[i - 1] = sumsOfAmountByImportanceAdjusted[i - 1] / entryAmounts[i - 1];
+                        averagesOfAmountByImportanceDefault[i - 1] = sumsOfAmountByImportanceDefault[i - 1] / entryAmounts[i - 1];
```
Comment (Stokalas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs:63]: Wait does these arrays just hold the same value or something like that? (the what's the point of having two) And if same operations in two lines, maybe lambda expression could be used here, but not sure, since it is pretty simple equation.

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 4]
```diff
@@ -8,42 +9,39 @@ namespace ePiggyWeb.DataManagement.Saving
 {
     public static class AlternativeSavingCalculator
```
Comment (Stokalas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs:9]: We should really think about leaving this class static of not.

**Review by Stokalas — COMMENTED** [review]
Since I was responsible for UI I still reviewed back-end code, and left few notices. @DeividasLT5

File: src/Pages/SavingSuggestions.cshtml.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            SetData();
+            return Page();
+        }
+
+        public void SetData()
+        {
+            UserId = int.Parse(User.FindFirst(ClaimTypes.Name).Value);
+            var dataManager = new DataManager(UserId);
+            Expenses = dataManager.Expenses.EntryList.GetFrom(StartDate).GetTo(EndDate);
+            Goal = dataManager.Goals.GoalList.FirstOrDefault(x => x.Id == Id);
```
Comment (ciliakas) [inline src/Pages/SavingSuggestions.cshtml.cs:67]: We can add DI to this class when we pull our other pull requests

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs (hunk 4 above)
Comment (ciliakas) [reply src/DataManagement/Saving/AlternativeSavingCalculator.cs:9]: We should ask the lecturer what he thinks, and where should we use static classes in general.

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 6]
```diff
@@ -8,42 +9,39 @@ namespace ePiggyWeb.DataManagement.Saving
...
     {
-        //This might not be 1M in the future, that why we are keeping this
-        private static decimal SavingRatio { get; } = 1M;
         private static decimal RegularSavingValue { get; } = 0.25M;
         private static decimal MaximalSavingValue { get; } = 0.5M;
         private static decimal MinimalSavingValue { get; } = 0.1M;
 
-        public static void GetSuggestedExpensesOffers(IEntryList entryList, IGoal goal, IList<ISavingSuggestion> entrySuggestions,
+        public static int GetSuggestedExpensesOffers(IEntryList entryList, IGoal goal, IList<ISavingSuggestion> entrySuggestions,
             List<SavingSuggestionByMonth> monthlySuggestions, decimal startingBalance, SavingType savingType = SavingType.Regular)
```
Comment (ciliakas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs:16]: To keep to our standard, maybe we should create an interface of SavingSuggestionByMonth, would also make this class less tightly coupled

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs (hunk 2 above)
Comment (ciliakas) [reply src/DataManagement/Saving/AlternativeSavingCalculator.cs]: actually just var, var[] would not compile :D

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 7]
```diff
@@ -8,42 +9,39 @@ namespace ePiggyWeb.DataManagement.Saving
...
                 {
                     var amountAfterSaving = savingType switch
                     {
-                        SavingType.Minimal => entry.Amount * SavingRatio * ratio * MinimalSavingValue,
-                        SavingType.Regular => entry.Amount * SavingRatio * ratio * RegularSavingValue,
-                        SavingType.Maximal => entry.Amount * SavingRatio * ratio * MaximalSavingValue,
+                        SavingType.Minimal => entry.Amount * ratio * MinimalSavingValue,
+                        SavingType.Regular => entry.Amount * ratio * RegularSavingValue,
+                        SavingType.Maximal => entry.Amount * ratio * MaximalSavingValue,
                         _ => throw new ArgumentOutOfRangeException()
```
Comment (ciliakas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs:44]: Even though this exception can't really occur ( because SavingType will always be one of the three values we have), we probably should still catch this exception

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 8]
```diff
@@ -8,42 +9,39 @@ namespace ePiggyWeb.DataManagement.Saving
...
+                var expenses = entryList.GetBy((Importance)i);
                 var ratio = enumCount - i;
                 foreach (var entry in expenses)
                 {
                     var amountAfterSaving = savingType switch
                     {
-                        SavingType.Minimal => entry.Amount * SavingRatio * ratio * MinimalSavingValue,
-                        SavingType.Regular => entry.Amount * SavingRatio * ratio * RegularSavingValue,
-                        SavingType.Maximal => entry.Amount * SavingRatio * ratio * MaximalSavingValue,
+                        SavingType.Minimal => entry.Amount * ratio * MinimalSavingValue,
```
Comment (ciliakas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs]: This whole methods is pretty large and quite hard to follow as it contains complex logic, it should probably by split into a few smaller methods if possible

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs [hunk 9]
```diff
@@ -54,19 +52,32 @@ public static class AlternativeSavingCalculator
...
+                        averagesOfAmountByImportanceDefault[i - 1] = 0;
+                    }
 
                     approximateSavedAmount += averagesOfAmountByImportanceDefault[i - 1] - averagesOfAmountByImportanceAdjusted[i - 1];
-
-                    monthlySuggestions.Add(new SavingSuggestionByMonth(averagesOfAmountByImportanceAdjusted[i - 1], averagesOfAmountByImportanceDefault[i - 1], (Importance)i));
-                }                              
+                }
+                timesToRepeatSaving++;              
+            }
```
Comment (ciliakas) [inline src/DataManagement/Saving/AlternativeSavingCalculator.cs:74]: For example this looks like it could be it's own separate method

**Review by ciliakas — CHANGES_REQUESTED** [review]
Seems to work after testing, but requires style adjustment and splitting into methods to make this quite complex logic easier to follow. Would be nice to maybe implement a requirement or two into this, I think it should be possible to add delegates or lambda expressions

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs (hunk 2 above)
Comment (Stokalas) [reply src/DataManagement/Saving/AlternativeSavingCalculator.cs]: Whoops, my bad :D

**Comment by Stokalas** [comment]
Since it is merged into calculations branch and this was mainly front-end branch merging pull-request and expecting mentioned fixes in calculations branch.

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs (hunk 3 above)
Comment (DeividasLT5) [reply src/DataManagement/Saving/AlternativeSavingCalculator.cs:63]: I'll look into lambda expressions, but the arrays just hold the untouched values, values after they're adjusted and how many values there are in an importance, there's probably a better way to do this, but I went for simplicity and readability. Answer your question?

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs (hunk 8 above)
Comment (DeividasLT5) [reply src/DataManagement/Saving/AlternativeSavingCalculator.cs]: Essentially the whole calculation works with two big cycles, I can't really imagine a way to split those up in a way that results in LESS confusing logic, because you'd have to trace what values and how are they worked with in different methods. I would consider it if you provided a better example of what could be done

File: src/DataManagement/Saving/AlternativeSavingCalculator.cs (hunk 9 above)
Comment (DeividasLT5) [reply src/DataManagement/Saving/AlternativeSavingCalculator.cs:74]: the problem is that the front would then need to call two seperate methods, no? where one is just a part of the whole, so I'm not sure about it. Or just seperate the method out and still call it, which, i feel like only adds to unreadability in this case.
